### PR TITLE
IR-267: add details for submanifest to the ImageStreamTag

### PR DESCRIPTION
**Pull request by prb112**
- Extends [image-stream-tag.tsx](https://github.com/openshift/console/blob/master/frontend/public/components/image-stream-tag.tsx) `https://<server>/k8s/ns/<namespace>/imagestreamtags/<imagestream>%3A<tag>` with submanifest details as a table
- The table includes dockerImageManifests details OS, Architecture, Identifier
- Matches the formatting of the details pane on the image-stream.tsx 
- The `SupportedPlatformTable` and pane are only presented when `tag.importPolicy.importMode` == `PreserveOriginal` per [OCPBUGS-7498](https://issues.redhat.com/browse/OCPBUGS-7498)
- The table is sorted using a lexigraphical order of OS, Architecture (and is sortable using the VolumesTable style of table)
- For consistency, Identifier is used instead of Digest

There was a consideration for the Identifier to be clickable, however there wasn't a clear destination for the next-page. 

I am unsure if there is a translation need on the word Manifest List on the new pane. I can adapt to the reviewers suggestion.

This work was done under the direction of @flavianmissi and @dmage and @dorzel 

An example of how it looks: 

image

A finer view of the detail:

image

With 'Legacy' or no list, then there is no Supported Platform displayed

image

Signed-off-by: Paul Bastide <pbastide@redhat.com>

### CODE REVIEW

File: frontend/public/components/image-stream-tag.tsx [hunk 1]
```diff
@@ -141,6 +148,39 @@ export const ImageStreamTagsDetails: React.SFC<ImageStreamTagsDetailsProps> = ({
           </div>
         )}
       </div>
+      <div className="co-m-pane__body-group">
+        <SectionHeading text={t('public~Manifest List')} />
```
Comment (flavianmissi) [inline frontend/public/components/image-stream-tag.tsx:152]: What do you think of "Supported Platforms" as a section header instead of "Manifest List'?

Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:152]: I like it. Honestly Manifest List seemed 'technical' and not descriptive.

Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:152]: I will change the text, and update as I fix the tests

Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:152]: I've made the change locally, I'm now working through test failures.

**Comment by prb112** [comment]
I looked at the failure image from the PR history [link](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/12549/pull-ci-openshift-console-master-e2e-gcp-console/1623489721348919296/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/e2e/add-flow-ci.feature/) for `e2e/add-flow-ci.feature` and it wasn't clear what failed or how related to the current PR. I am waiting on the PR checks/tests to finish.

**Comment by flavianmissi** [comment]
last failure was different than the previous. will retry once more.
/retest

**Comment by prb112** [comment]
> last failure was different than the previous. will retry once more. /retest

Thank you Flavian, I was going through the image captures, and didn't see any of the pages I changed presented in the logs/images/videos.

**Comment by prb112** [comment]
> last failure was different than the previous. will retry once more. /retest

rebased hoping this helps

**Comment by prb112** [comment]
Note sure why this page should fail - [Create the different workloads from Add page -- Create Sample Application from Add page GS-03-TC05 (example #2) (failed).png](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/12549/pull-ci-openshift-console-master-e2e-gcp-console/1624095039049175040/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/e2e/add-flow-ci.feature/Create%20the%20different%20workloads%20from%20Add%20page%20--%20Create%20Sample%20Application%20from%20Add%20page%20GS-03-TC05%20(example%20#2)%20(failed).png)

/test e2e-gcp-console

**Comment by prb112** [comment]
```
   * could not run steps: step e2e-gcp-console failed: "e2e-gcp-console" pre steps failed: "e2e-gcp-console" pod "e2e-gcp-console-ipi-install-install" failed: the pod ci-op-l62pgc9p/e2e-gcp-console-ipi-install-install failed after 1h0m51s (failed containers: test): ContainerFailed one or more containers exited 
```

seems unrelated

/test e2e-gcp-console

File: frontend/public/components/image-stream-tag.tsx [hunk 2]
```diff
@@ -141,6 +148,39 @@ export const ImageStreamTagsDetails: React.SFC<ImageStreamTagsDetailsProps> = ({
           </div>
         )}
       </div>
+      <div className="co-m-pane__body-group">
+        <SectionHeading text={t('public~Supported Platforms')} />
```
Comment (jhadvig) [inline frontend/public/components/image-stream-tag.tsx:152]: So this approach wont enable column based sorting of the table.
Please check how other tables are implemented, for example how [pod details page](https://github.com/openshift/console/blob/master/frontend/public/components/pod.tsx#L857) is implementing [VolumesTable](https://github.com/openshift/console/blob/master/frontend/public/components/volumes-table.tsx#L167)

File: frontend/public/components/image-stream-tag.tsx [hunk 3]
```diff
@@ -141,6 +148,39 @@ export const ImageStreamTagsDetails: React.SFC<ImageStreamTagsDetailsProps> = ({
...
+        {_.isEmpty(sortedSubmanifests) && importPolicyPreserveOriginal ? (
+          <span className="text-muted">{t('public~No Supported Platforms')}</span>
```
Comment (jhadvig) [inline frontend/public/components/image-stream-tag.tsx:154]: Please use the `EmptyBox` component here

**Review by jhadvig — CHANGES_REQUESTED** [review]

File: frontend/public/components/image-stream-tag.tsx (hunk 2 above)
Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:152]: Thanks, I presorted the columns - I can add per citation. thank you

File: frontend/public/components/image-stream-tag.tsx (hunk 3 above)
Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:154]: Will do - thanks

Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx:154]: Resolved locally - it is in my updated commit.

File: frontend/public/components/image-stream-tag.tsx (hunk 2 above)
Comment (prb112) [reply frontend/public/components/image-stream-tag.tsx]: I've updated based on VolumesTable's style/code

**Comment by prb112** [comment]
Hi @jhadvig I have addressed your comments, and look forward to your further feedback.  Thank you, Paul

**Comment by xiuwang** [comment]
@prb112  I raised a bug https://issues.redhat.com/browse/OCPBUGS-7498, please help have a look, thanks

**Comment by prb112** [comment]
There was a timeout in one of the tests

**Comment by xiuwang** [comment]
@prb112 Could you check[ my comment in bug](https://issues.redhat.com/browse/OCPBUGS-7498?focusedCommentId=21756091&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-21756091)?

**Comment by prb112** [comment]
Followed up in the Bug. 

There was a timeout in the tests 
/test e2e-gcp-console

**Comment by prb112** [comment]
I reviewed the failures. 

```
1) Create the different workloads from Add page
2) Create the different workloads from Add page
       Create Sample Application from Add page: GS-03-TC05 (example #2):
     CypressError: Timed out retrying after 40050ms: `cy.click()` failed because this element is detached from the DOM.
```

I checked... https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_console/12549/pull-ci-openshift-console-master-e2e-gcp-console/1626543238082990080/artifacts/e2e-gcp-console/test/artifacts/gui_test_screenshots/cypress/screenshots/e2e/add-flow-ci.feature/ 

These pages do not overlap with my changes, and it's not clear that there is an issue.
Retesting

/test e2e-gcp-console

**Comment by prb112** [comment]
I looked at the video and screenshots for the test - the Samples page seems to timeout during the final steps. I'm waiting for the PR to be marked failed, and then I'll retest

image

**Comment by prb112** [comment]
restarting
/test e2e-gcp-console

**Comment by prb112** [comment]
Hey @xiuwang the PR passed. Let me know if there are any issues.

**Comment by prb112** [comment]
Hi @flavianmissi I've updated, if you can review, thank you

**Comment by prb112** [comment]
Thank you @xiuwang

File: frontend/public/components/image-stream-tag.tsx [hunk 4]
```diff
@@ -34,6 +35,90 @@ const splitEnv = (nameValue: string) => {
...
+      sortField: 'digest',
+      transforms: [sortable],
+      props: { className: supportedPlatformColumnClasses[2] },
+    },
+  ];
+
+  if (!policy || submanifests.length === 0) {
+    // If the policy does not support Manifest Lists, it exits.
+    // or the Manifest List is empty
+    return null;
```
Comment (jhadvig) [inline frontend/public/components/image-stream-tag.tsx:98]: if there are no `submanifests` we should return `EmptyBox` component.
Check howe its done for [VolumesTable](https://github.com/openshift/console/blob/master/frontend/public/components/volumes-table.tsx#L213-L232)

**Pull request merged**